Types/ScreenHandler.cs

```diff
@@ -0,0 +1,127 @@
+ * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
+ * IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
+ * CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
+ * TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
+ * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+ */
+
+using System.Collections.Generic;
+
+namespace SteelSeries {
+
+    namespace GameSense {
+
+        [FullSerializer.fsObject(Converter = typeof(ScreenHandlerConverter))]
+        [UnityEngine.CreateAssetMenu(fileName = "ScreenHandler", menuName = "GameSense/Handlers/Screen Handler")]
+        [System.Serializable] public class ScreenHandler : AbstractHandler {
+            private ScreenDataType _screenDataType { get; set; }
+            public ScreenDataType screenDataType {
+                get { return _screenDataType; }
+            }
+
+            // device - zone
+            public DeviceZone.AbstractScreenDevice_Zone deviceZone;
+
+            // mode
+            public ScreenMode mode;
+
+            // datas
+            public AbstractScreenData datas;
+            public ScreenDataStatic data_static {
+                get { return datas as ScreenDataStatic; }
+                set { datas = value; _screenDataType = ScreenDataType.Static; }
+            }
+
+            public ScreenDataRanges data_ranges {
+                get { return datas as ScreenDataRanges; }
+                set { datas = value; _screenDataType = ScreenDataType.Range; }
+            }
+
+            private static ScreenHandler _new() {
+                ScreenHandler sh = CreateInstance< ScreenHandler >();
+                return sh;
+            }
+
+            private static ScreenHandler Create( DeviceZone.AbstractScreenDevice_Zone dz, ScreenMode mode ) {
+                ScreenHandler sh = _new();
+                sh.deviceZone = dz;
+                sh.mode = mode;
+                return sh;
+            }
+
+            public static ScreenHandler Create( DeviceZone.AbstractScreenDevice_Zone dz, ScreenMode mode, AbstractFrameData[] datas ) {
+                ScreenHandler sh = Create( dz, mode );
+                sh.data_static = ScreenDataStatic.Create( datas );
+                return sh;
+            }
+
+            public static ScreenHandler Create( DeviceZone.AbstractScreenDevice_Zone dz, ScreenMode mode, FrameDataRange[] datas ) {
+                ScreenHandler sh = Create( dz, mode );
+                sh.data_ranges = ScreenDataRanges.Create( datas );
+                return sh;
+            }
+
+            /// <summary>
+            /// Invoke Preprocess for each contained frame data.
+            /// Needs to be called on the main thread (from Monobehaviour script).
+            /// </summary>
+            public override void Preprocess() {
+                switch ( screenDataType ) {
+
+                    case ScreenDataType.Static:
```

Comment:
To anyone who gets a notification about this comment: Hi - we're looking at merging this soon.
I'm seeing a bug where when you assign ranged screen data to the screen handler in the inspector, the handler's _screenDataType isn't being set to ScreenDataType.Ranged, and instead defaults to static. This causes bad behavior when the event is triggered.